app/scripts/lib/channels/mixins/postmessage_receiver.js

```diff
@@ -88,9 +91,35 @@ define([
       setResponseTimeoutTimer.call(this, outstanding);
     },
 
+    isEventFromExpectedOrigin: function (event) {
+      // `message` events that come from the Fx Desktop browser have an
+      // origin of the string constant 'null'. See
+      // https://developer.mozilla.org/docs/Web/API/Window/postMessage#Using_win.postMessage_in_extensions
+      // and
+      // https://bugzilla.mozilla.org/show_bug.cgi?id=1040257
+      //
+      // Events from the browser are trusted by default.
+      if (event.origin === 'null') {
+        return true;
+      }
+
+      // If the event is not from the browser, it must be from
+      // the expected origin.
+      return this._origin === event.origin;
+    },
+
     receiveMessage: function (event) {
-      // TODO - Ensure the event.origin is one we trust,
-      // esp in the iframe case
+      if (! this.isEventFromExpectedOrigin(event)) {
+        // from an unexpected origin, drop it on the ground.
+        var err = AuthErrors.toError('UNEXPECTED_POSTMESSAGE_ORIGIN');
+
+        // set the unexpected origin as the context, this will be logged.
+        err.context = event.origin;
+
+        this.window.console.error('postMessage received from unexpected origin');
```

Comment:
nit: we could log the actual and expected origins here as well, to assist debugging